### ISSUE TITLE
[RFC]: Make any vLLM model a pooling model

### ISSUE THREAD

**Issue by DarkLight1337** [vllm-project/vllm#10674]
### Motivation.

Currently, we have to open new PRs to add pooling functionality for existing architectures supported in vLLM. Since the code involved is basically the same for each model, there is potential to automate away this boilerplate.

### Proposed Change.

Implement a pooling adapter that can be applied to any existing text generation model in vLLM. To preserve features such as LoRA, PP and multimodality, the adapter simply creates a new subclass of the original model.

- [x] Embedding model adapter: #10769
- [ ] Classification model adapter #11469
- [ ] Reward model adapter #11469

The pooling adapter to apply depends on the purpose of the model. To facilitate this, the `embedding` task will be split up, so that the user can specify which adapter to apply to the model via `--task`:

- [x] #10820
- [x] #11457

Meanwhile, current embedding-related classes will be renamed to avoid confusion between `embed` and other pooling tasks:

- [x] #10801

### Feedback Period.

1-2 weeks

### CC List.

@youkaichao @mgoin @robertgshaw2-neuralmagic @maxdebayser

### Any Other Things.

Note that we can still directly map to pooling models in the model registry. This is used when the model architecture has different pooling defaults (e.g. `pooling_type=CLS` for BERT) or additional modules (e.g. `score` in `Qwen2ForRewardModel`). For models that already support pooling, the adapter returns the original model without modifications.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by maxdebayser** [comment]
What would happen if the user loads a BertModel without `--task-embedding`? Would there be a validation to prevent this from happening our would we allow the decoder-only model runner to try to run the model for text generation?

**Comment by DarkLight1337** [comment]
In `is_text_generation_model`, we check for the existence of `compute_logits` and `sample` methods. Since those methods aren't in `BertModel`, `--task generation` is not allowed for that model.

**Comment by maxdebayser** [comment]
Ok, I think this makes sense.

**Comment by DarkLight1337** [comment]
Update: I have expanded the scope of this RFC to cover all pooling-based models.
